BUG: Distinguish exact vs. equivalent dtype for C type aliases.

eirrgang opened the pull request:
For `asarray` and for the `dtype` equality operator,
equivalent dtype aliases were considered exact matches.
This change ensures that the returned array has a descriptor
that exactly matches the requested dtype.

Note: Intended behavior of `np.dtype('i') == np.dtype('l')`
is to test compatibility, not identity. This change does not
affect the behavior of `PyArray_EquivTypes()`, and the
`__eq__` operator for `dtype` continues to map to
`PyArray_EquivTypes()`.

Fixes #1468.

eirrgang committed:
BUG: Distinguish exact vs. equivalent dtype for C type aliases.

For `asarray` and for the `dtype` equality operator,
equivalent dtype aliases were considered exact matches.
This change ensures that the returned array has a descriptor
that exactly matches the requested dtype.

Note: Intended behavior of `np.dtype('i') == np.dtype('l')`
is to test compatibility, not identity. This change does not
affect the behavior of `PyArray_EquivTypes()`, and the
`__eq__` operator for `dtype` continues to map to
`PyArray_EquivTypes()`.

Fixes #1468.

eirrgang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1627,12 +1627,30 @@ _array_fromobject_generic(
+                            PyArray_DIMS(oparr),
+                            PyArray_STRIDES(oparr),
+                            PyArray_DATA(oparr),
+                            PyArray_FLAGS(oparr),
```
or should this be `type->flags`?

seberg replied:
This is right, I guess you don't even need it, the only one would be writeonly, and I think we check the base one for that.  But it doesn't hurt (the owndata flag is removed automatically).

eirrgang committed:
Use the correct pointer type.

eirrgang committed:
Coerce the returned pointer type.

eirrgang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1627,12 +1627,30 @@ _array_fromobject_generic(
             if (copy != NPY_COPY_ALWAYS && STRIDING_OK(oparr, order)) {
                 Py_INCREF(op);
-                ret = oparr;
+                if (oldtype == type) {
```
It seems like either
* this operator is overly strict (or the wrong check), or
* this change fundamentally conflicts with the current notions of `is`

eirrgang replied:
Maybe this change is too deep. If a user says `asarray(a, dtype='i')`, are there members of https://numpy.org/devdocs/reference/c-api/types-and-structures.html#c.PyArray_Descr that might differ under ordinary circumstances? Might `hash` be uninitialized?

seberg replied:
I think this is right, and we just need to rethink the tests that are failing!

eirrgang committed:
Get the base object.

Use the correct API call to actually set the base object.

eirrgang committed:
Add unit testing.

eirrgang committed:
Don't regenerate the descriptor unnecessarily.

Fix a false mismatch. Separate dtype objects, even if equivalent,
cause distinct array views to be created.

eirrgang committed:
Update comment and obey formatting requirements.

eirrgang commented:
The updates that were necessary for the tests indicate that this change will surprise some users. Some sort of release note is warranted, and perhaps some sort of update to docs related to the `dtype` kwarg.

Suggestions for specific docs files to update?

eric-wieser commented on numpy/array_api/tests/test_asarray.py:
```diff
@@ -0,0 +1,24 @@
+    assert np.asarray(a) is a
+    assert np.asarray(a, dtype='i') is a
+    # This may produce a new view or a copy, but is never the same object.
+    assert np.asarray(a, dtype='l') is not a
```
Note that this is only a good test on windows; it's `l` and `q` that clash on Linux, where l`long` is 64 bit

eirrgang replied:
Good point. I have added more permutations to this block so that at least one check would not pass without this patch.

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -5387,7 +5387,7 @@ def test_ufunc_with_out_varied():
 
 def test_astype_mask_ordering():
-    descr = [('v', int, 3), ('x', [('y', float)])]
+    descr = np.dtype([('v', int, 3), ('x', [('y', float)])])
```
So record dtypes are now nominal not structural? That is, two record types with identical fields are treated differently depending on whether they are at the same memory locations? That seems undesirable to me; although if structured dtypes have mutable state maybe it's the right choice after all.

Either way, it would be good to test this explicitly rather than only accidentally as part of this test

seberg replied:
Well, we can either preserve the array identity check faithfully, or the dtype identity check one, since doing `np.dtype(something)` will give a new dtype each time.

Mutibility is a point (although I don't like it for dtypes).  I guess the structured dtype case is "more equivalent" than for the ones with differing type numbers...
But is there much point in preserving the array identity?

eirrgang replied:
I renamed the new test function to make it (hopefully) more obvious that it includes tests of the new dtype equivalence/identity behavior. Does those assertions seem sufficient?

eirrgang replied:
I think the latest test is pretty thorough and illustrative. If a different test file name would help to find the expression of these identity semantics, please advise.

eirrgang committed:
Expand test_asarray.py.

* Improve comments/docs.
* Improve descriptiveness of variable names.
* Add additional test expressions that would not pass without
  this patch.

eirrgang committed:
Lint.

Shorten some lines.

eirrgang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1627,12 +1627,31 @@ _array_fromobject_generic(
             }
         }
-        /* One more chance */
+        /* One more chance for faster exit if user specified the dtype. */
```
It seems like the whole conditional (starting at 1612) might be appropriately replaced with a call to `array_astype()`, except that it looks like there are some default behaviors that are different (e.g. forcecopy).

In particular, the use of `PyArray_DescrConverter -> PyArray_AdaptDescriptorToArray` in `methods.c` `array_astype()` has slightly different behavior than the `PyArray_DescrConverter2` -> `PyArray_DescrConverter` here.

It seems like a separate issue, but maybe any associated behavior changes are something that should be considered before release notes are finalized?

seberg replied:
Its a good point that it may be nice to consolidate the fast-path with `astype`.  But I think its OK to not do it for now.

eirrgang commented on numpy/array_api/tests/test_asarray.py:
```diff
@@ -0,0 +1,45 @@
+    assert np.asarray(int_array, dtype='q') is not int_array
+    assert np.asarray(long_int_array, dtype='q') is not long_int_array
+    assert long_int_array is not np.asarray(int_array, dtype='l')
+    assert long_int_array.base is np.asarray(int_array, dtype='l').base
```
Oops. This won't be true on 32-bit systems, right?

eirrgang committed:
Add release note and further clarify tests.

eirrgang commented:
> The updates that were necessary for the tests indicate that this change will surprise some users. Some sort of release note is warranted, and perhaps some sort of update to docs related to the `dtype` kwarg.
> 
> Suggestions for specific docs files to update?

I couldn't find any docs that were specific enough to require updating, but I added a release notes doc.

seberg commented:
Thanks this is awesome.  Those are long docs, if we know a place to put them, we probably should... For the release note, it would be nice to make it much shorter unfortuntely!

eirrgang commented:
> Thanks this is awesome. Those are long docs, if we know a place to put them, we probably should... For the release note, it would be nice to make it much shorter unfortuntely!
> 
> EDIT: I (or someone else from us) can take a stab at that though!

Better now?

eirrgang commented on doc/release/upcoming_changes/21995.compatibility.rst:
```diff
@@ -0,0 +1,46 @@
+If the provided dtype is compatible, but not identically the same
+:py:class:`dtype` object, a new array handle is always created with
+a reference to the user-provided dtype instance.
+If the data type is compatible, and copying is not required, the new
+`ndarray` uses the original array as its
+`base <https://numpy.org/doc/stable/reference/generated/numpy.ndarray.base.html>`__.
+
+Before this change, for two equivalent but non-identical dtypes,
+
+    assert isinstance(typeA, np.dtype) and isinstance(typeB, np.dtype)
+    assert typeA == typeB
+    assert typeA is not typeB
+    if my_array.dtype is typeA:
+        assert my_array is np.asarray(my_array, dtype=typeB)
+        assert np.asarray(my_array, dtype=typeB).dtype is not typeB
+
+This change allows programs to be able to reliably get the exact dtype
+representation they request, regardless of possibly aliased types on the
+calling platform.
+
+However, identity semantics for array results and their
+dtype members may require minor updates to calling code.
+
+After this change, on a system where C ``int`` and C ``long`` are the same
+precision, ``np.dtype('i') == np.dtype('l')``,
+but ``np.dtype('i') is not np.dtype('l')``.
+
+    assert int_array.dtype is np.dtype('i')
+    long_int_array = np.asarray(int_array, dtype='l')
+    assert long_int_array is not int_array
+    if np.dtype('i') == np.dtype('l'):
+        assert int_array is long_int_array.base
+
+New array views are created with each call to `asarray` with non-identical
+dtype kwarg, but the underlying data is the same.
+
+    assert int_array.dtype is np.dtype('i')
+    long_int_array = np.asarray(int_array, dtype='l')
+    assert long_int_array is not np.asarray(int_array, dtype='l')
+    assert long_int_array.base is np.asarray(int_array, dtype='l').base
```
or replace third paragraph with the following summary:

```
* Resulting array *dtype* members are now guaranteed to pass an identity check (``is``) with the caller-provided dtype.
* Resulting arrays may no longer pass an identity check with the provided *obj* if the provided dtype is compatible, but not identical. However, if copy was avoided, ``result.base is obj``.
```

seberg replied:
Sorry, I was out for a while and lost track of this.  I pushed an alternative based on your suggestions (opted for an example anyway though).  Would love it shorter, but I hope it is fairly concise.
Maybe you can have a quick look @eirrgang, if you don't like it happy if you revert or suggest changes.

eirrgang commented:
> Thanks this is awesome. Those are long docs, if we know a place to put them, we probably should... For the release note, it would be nice to make it much shorter unfortuntely!
> 
> EDIT: I (or someone else from us) can take a stab at that though!

Is my proposed edit better? Or is there an alternative proposal?

seberg committed:
DOC: Take a stab at shortening the release note

mattip commented:
Thanks @eirrgang

eirrgang commented on doc/release/upcoming_changes/21995.compatibility.rst:
```diff
@@ -1,46 +1,21 @@
+    >>> arr = np.array([1, 2, 3], dtype="long")
+    >>> new_dtype = np.dtype("longlong")
+    >>> new = np.asarray(arr, dtype=new_dtype)
+    >>> new.dtype is dtype
```
Was this supposed to be `new.dtype is new_dtype`?

mattip replied:
yes

eirrgang reviewed (commented):
Sorry. I was away for the weekend and just got caught up. Looks good, but I think there is a typo.

mattip commented:
See #22227

InessaPawson commented:
Hi-five on merging your first pull request to NumPy, @eirrgang! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion

rgommers commented:
As reported in gh-22233, there is a serious issue with this PR, likely a memory leak that is causing issues for downstream projects like SciPy and MNE-Python.

I will revert this PR. @eirrgang it would be great if you could resubmit this as a new PR, and then we can merge it again once the problem is understood and addressed. Maybe `pytest-leaks` or Valgrind will be helpful here in pointing to the root cause.

rgommers commented:
Ugh, GitHub: _Sorry, this pull request couldn’t be reverted automatically. It may have already been reverted, or the content may have changed since it was merged._

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1627,12 +1627,31 @@ _array_fromobject_generic(
+                            PyArray_FLAGS(oparr),
+                            op,
+                            op
+                            );
```
Oh, either I just didn't think of it, or expected that this steals a reference to `op` for the `base`.  We are missing a DECREF, I will make a PR.

eirrgang replied:
Yes, I think I originally tried `PyArray_NewFromDescr` with `op` for `*data` and `nullptr` for `*obj`, and `PyArray_SetBaseObject` or something that steals the reference for the base. Anyway, I guess we were sloppy when switching to `PyArray_NewFromDescrAndBase`. Thanks for the quick catch!

leofang commented:
It is incorrect to add `numpy/array_api/tests/test_asarray.py`. The test itself is not of concern to Array API. It should be moved elsewhere. 

(This is why we need to set up `array_api` namespace owners... cc: @seberg @rgommers)

seberg commented:
Yes, sorry, I missed where this test was, it should be elsewhere.  I had just thought it looked reasonably thorough.  It might also be split up, but I don't care.

In any case, we need to move this test into `/core/tests/test_<something>` not here. @eirrgang do you want to have a look into that.

eirrgang commented:
> Yes, sorry, I missed where this test was, it should be elsewhere. I had just thought it looked reasonably thorough. It might also be split up, but I don't care.
> 
> In any case, we need to move this test into `/core/tests/test_<something>` not here. @eirrgang do you want to have a look into that.

Sure. I'll make a new PR. As I recall, there wasn't an obvious good place for the test. I'll look again, and try to come up with something reasonable.

seberg commented:
Way too long, but `test_multiarray.py` with `TestCreation` might be an option.  Otherwise `test_array_coercion.py` has somewhat related tests so you could just add it in that file.  The only real problem here is that it doesn't belong under `array_api/tests`, but `core/tests`

WarrenWeckesser commented:
This appears to have broken some tests in `scipy.sparse`.  Ultimately, the problem comes down to the dtype check requiring that the dtypes are the *same object*, instead of being equal in the Python sense.

This shows the essence of the problem:
```
In [27]: a = np.array([10, 20, 30, 40])

In [28]: b = np.asarray(a, dtype=a.dtype.newbyteorder('native'))

In [29]: a is b  # Before this PR, this would be true.
Out[29]: False

In [30]: a.dtype == b.dtype
Out[30]: True
```
The dtypes of `a` and `b` are equal, but `asarray()` still created a view.  There are tests in `scipy.sparse` that expect `a is b` to be True.

eirrgang commented:
> `a.dtype.newbyteorder('native')` creates a new dtype instance, but the new dtype is equal to (i.e. `==`) `a.dtype`. So the dtypes of `a` and `b` are equal, but `asarray()` still created a view. There are tests in `scipy.sparse` that expect `a is b` to be True.

Yes. My understanding from @seberg is that this is the intended behavior, and it is reflected in the updates to the numpy tests and release notes. Please advise if you find contradictions in the documentation. Some documentation was trimmed, so you might find the commit history of the PR informative. See also https://github.com/numpy/numpy/pull/21995/files#r922726648

seberg commented:
@WarrenWeckesser, yeah, that was intentional but that doesn't mean it was the best thing to do.  Please do followup if you think we should consider undoing (parts) of this.
